subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/TextResourceCodec.kt

```diff
@@ -0,0 +1,104 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.instantexecution.serialization.codecs
+
+import org.gradle.api.internal.file.FileOperations
+import org.gradle.api.internal.resources.ApiTextResourceAdapter
+import org.gradle.api.internal.resources.ClasspathBackedTextResource
+import org.gradle.api.internal.resources.FileCollectionBackedArchiveTextResource
+import org.gradle.api.internal.resources.FileCollectionBackedTextResource
+import org.gradle.api.internal.resources.StringBackedTextResource
+import org.gradle.api.resources.TextResource
+import org.gradle.api.resources.TextResourceFactory
+import org.gradle.instantexecution.extensions.uncheckedCast
+import org.gradle.instantexecution.serialization.Codec
+import org.gradle.instantexecution.serialization.ReadContext
+import org.gradle.instantexecution.serialization.WriteContext
+import org.gradle.instantexecution.serialization.ownerService
+import org.gradle.instantexecution.serialization.readFile
+import org.gradle.instantexecution.serialization.writeFile
+import java.nio.charset.Charset
+
+
+object TextResourceCodec : Codec<TextResource> {
+
+    override suspend fun WriteContext.encode(value: TextResource) {
+        when (value) {
+            is ApiTextResourceAdapter -> {
+                writeSmallInt(API_ADAPTER)
+                writeString(value.uri.toASCIIString())
+            }
+            is FileCollectionBackedArchiveTextResource -> {
+                writeSmallInt(ARCHIVE)
+                writeFile(value.archiveFile)
+                writeString(value.path)
+                writeString(value.charset.name())
+            }
+            is FileCollectionBackedTextResource -> {
+                writeSmallInt(FILE_COLLECTION)
+                writeFile(value.inputFiles.singleFile)
+                writeString(value.charset.name())
+            }
+            is ClasspathBackedTextResource -> {
+                writeSmallInt(CLASSPATH)
+                writeClass(value.baseClass)
+                writeString(value.path)
+                writeString(value.charset.name())
+            }
+            is StringBackedTextResource -> {
+                writeSmallInt(STRING)
+                writeString(value.asString())
+            }
+            else -> {
+                writeSmallInt(UNKNOWN)
+                write(value)
+            }
+        }
+    }
+
+    override suspend fun ReadContext.decode(): TextResource =
+        when (readSmallInt()) {
+            API_ADAPTER -> textResourceFactory.fromInsecureUri(readString())
```

Comment:
I would think about decoding all of them as a fixed string or file resource, for now, just like we do for providers. This way, all of the resolution and decoding work is done when we write the cache, rather than on each build. Then, we can later fix the cases where the source of the resource is some changing thing.